### PR TITLE
Implement unshipped messages in message port.

### PR DESCRIPTION
**Pull request by jellefoks**
Implement storage of unshipped message in message port, to allow a message port to receive and hold messages before the destination is ready to receive them.

Use the above functionality to unblock worker.postMessage calls to no longer wait until the worker script completes loading and executing.

b/310240306
b/226640425

### CODE REVIEW

**Comment by jellefoks** [comment]
I should probably add a unit test, but to do that without becoming implementation specific I should not rely on `SetEventTarget` to enable delivery of messages. unshipped message probably instead should be held until a message event handler is registered. Done that way, the edge cases can probably be tested without relying on implementation.

**Comment by jellefoks** [comment]
> I should probably add a unit test, but to do that without becoming implementation specific I should not rely on `SetEventTarget` to enable delivery of messages. unshipped message probably instead should be held until a message event handler is registered. Done that way, the edge cases can probably be tested without relying on implementation.

Since MessagePort is not constructible from JS, that becomes more involved, but also unshipped_messages_ access and mutations have to be made thread safe.

File: cobalt/web/message_port.h [hunk 1]
```diff
@@ -91,15 +93,31 @@ class MessagePort : public script::Wrappable,
                                              event_listener);
   }
 
-  web::EventTarget* event_target() { return event_target_; }
+  EventTarget* event_target() const { return event_target_; }
+  Context* context() const {
+    return event_target_->environment_settings()->context();
```
Comment (sideb0ard) [inline cobalt/web/message_port.h:98]: do we need to check event_target_ exists before calling environment_settings()?

**Review by sideb0ard — APPROVED** [review]
minor comments, lgtm

File: cobalt/web/message_port.h (hunk 1 above)
Comment (jellefoks) [reply cobalt/web/message_port.h:98]: It's only used in code where that check is already done, but that may not always be the case in the future, so adding that check can be useful and the processing overhead is trivial in this case.

File: cobalt/web/message_port.h [hunk 2]
```diff
@@ -34,14 +36,16 @@ namespace cobalt {
 namespace web {
 
 class MessagePort : public script::Wrappable,
-                    public base::SupportsWeakPtr<MessagePort>,
                     public Context::EnvironmentSettingsChangeObserver {
  public:
-  explicit MessagePort(web::EventTarget* event_target);
-  ~MessagePort();
+  MessagePort() {}
```
[the remark this answers is not in the document]
Comment (jellefoks) [reply cobalt/web/message_port.h]: Thanks. Done.

File: cobalt/web/message_port.h (hunk 1 above)
Comment (jellefoks) [reply cobalt/web/message_port.h]: Thanks. Done.
